### ISSUE TITLE
Infer behaves differently with and without the loop initialization expression

### ISSUE THREAD

**Issue by zyh1121** [facebook/infer#1374]
Please make sure your issue is not addressed in the [FAQ](https://fbinfer.com/docs/support#troubleshooting).

Please include the following information:
- [x] The version of infer from `infer --version`.
  ```
  $ infer --version
  Infer version v1.0.0-e065b0b0b
  ```
- [x] Your operating system and version, for example "Debian 9", "MacOS High Sierra", whether you are using Docker, etc.
  ```
  Linux ... 3.10.0-957.el7.x86_64
  Red Hat Enterprise Linux Server release 7.6 (Maipo)
  ```
- [x] Which command you ran, for example `infer -- make`.
```
infer --bufferoverrun --pulse --bo-field-depth-limit 3 --enable-issue-type BUFFER_OVERRUN_L5  run -- gcc -c $src_file
```
- [x] The full output in a paste, for instance a [gist](https://gist.github.com/).
- [x] If possible, a minimal example to reproduce your problem (for instance, some code where infer reports incorrectly, together with the way you run infer to reproduce the incorrect report).

#### * no for-loop init expression --> not buggy

```c
$ cat 1.c 
main ( ) 
{ 
int entity_7 ; 
char entity_4 [ 48 ] ; 
entity_7 = 0 ; 
int entity_1 ; 
for ( ; ; ) { 
} 
entity_4 [ entity_1 ] 
; 
}
```

```shell
$ infer --bufferoverrun --pulse --enable-issue-type BUFFER_OVERRUN_L5  run -- gcc -c 1.c 
Capturing in make/cc mode...
Found 1 source file to analyze in ..../infer-out
1/1 [################################################################################] 100% 52.458ms

  No issues found  
```

#### * for-loop init expression --> buggy

```c
$ cat 2.c 
main ( ) 
{ 
int entity_7 ; 
char entity_4 [ 48 ] ; 
entity_7 = 0 ; 
int entity_1 ; 
for (entity_1 ; ; ) { 
} 
entity_4 [ entity_1 ] 
; 
}
```

```shell
$ infer --bufferoverrun --pulse --enable-issue-type BUFFER_OVERRUN_L5  run -- gcc -c 2.c 
Capturing in make/cc mode...
Found 1 source file to analyze in ..../infer-out
1/1 [################################################################################] 100% 30.304ms

2.c:9: error: Buffer Overrun L5
  Offset: [-oo, +oo] Size: 48.
   7. for (entity_1 ; ; ) { 
   8. } 
   9. entity_4 [ entity_1 ] 
      ^
  10. ; 
  11. }


Found 1 issue
            Issue Type(ISSUED_TYPE_ID): #
  Buffer Overrun L5(BUFFER_OVERRUN_L5): 1
```

----

In addition, any chance there is a way to **let infer ignore `entity_4 [ entity_1 ]`** in both cases because it's not reachable? We have a lot of such cases so it would be super helpful if we could do that. 

Thanks a lot!!

**Comment by rvantonder** [comment]
Hey @zyh1121 couldn't help but notice some of your reports on some syntax affecting analysis behavior. Another thing you could try, while waiting for a fix (or if it turns out to be difficult to fix all the cases you're running into) is to rewrite your program for these problematic cases. This assuming that you can convince yourself that changing braces/init expressions/unreachable code shouldn't in principle affect the overall result of course. I think this is the case of the examples you posted. I wrote a tool called [comby](https://comby.dev) to make some of these kinds of program changes easy (and also used it in the context of [getting around analyzer issues](https://rijnard.com/pdfs/tailoring-analysis-icse-2020.pdf)) so you may have some luck.

Example patterns to rewrite your program to remove the for init-expression:

```
int :[[entity]] ;
for (:[[entity]] ; ; )
```

-> 

```
int :[[entity]] ;
for ( ; ; )
```

[Full example in the playground](https://bit.ly/2XmwAjU).

Here's an example for erase all the statements after a for loop matching a particular syntax (e.g., the for-syntax that implies the rest of the block is unreachable):

`{ :[before] for (:[_]; ; ) {:[body]} :[after] }` -> `{ :[before] for (:[_] ; ; ) {:[body]} }`

[Full example in playground](https://bit.ly/2XmHIwV). 

These are just indicative examples--if you need more help with patterns let me know, there's a gitter channel, or just e-mail me.

**Comment by zyh1121** [comment]
@rvantonder Thanks for the comments! I could imagine the comby tool can help us and transform such programs to walkaround such syntax issues. Actually, we did some simple hack for the bracket issue before reporting. :)  comby would be handy and a systematic approach for such cases! 

However, in our scenario, most of such snippets are not written by human developers. They are synthesized and re-assembled based on real-world programs. Basically, we use infer as a judge to generate a lot of snippets with different patterns for model training. Before sending to infer, we make sure they are valid programs per clang/gcc. In general, it's a bit challenging to identify if the strange behaviors were caused by syntax or semantics. And we are not sure if such syntax issues could lead to other issues. So, we thought it might be a good idea to report such cases. 

In addition, as I mentioned in the question, we are also very interested in learning if there are some tricks we could play with the reachability.

Thanks again.
